connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java

```diff
@@ -256,39 +277,76 @@ public void testPartitionStateMutation() {
         partitionStates.put(sourceTopicPartition, partitionState);
         RecordMetadata recordMetadata = new RecordMetadata(sourceTopicPartition, metadataOffset, 0, 0, 0, recordPartition);
 
-        when(outstandingOffsetSyncs.tryAcquire()).thenReturn(true);
+        doAnswer(new Answer<Object>() {
+            @Override
+            public Object answer(final InvocationOnMock invocation) {
+                final Callback callback = invocation.getArgument(1);
+                callback.onCompletion(null, null);
+                return null;
+            }
+        }).when(producer).send(any(), any());
         mirrorSourceTask.commitRecord(sourceRecord, recordMetadata);
         assertEquals(recordOffset, partitionState.lastSyncUpstreamOffset, "sync offsets");
         assertEquals(metadataOffset, partitionState.lastSyncDownstreamOffset, "sync offsets");
         assertEquals(recordOffset, partitionState.previousUpstreamOffset, "sync offsets");
         assertEquals(metadataOffset, partitionState.previousDownstreamOffset, "sync offsets");
-        assertFalse(partitionState.shouldSyncOffsets);
+        assertFalse(partitionState.shouldSyncOffsets, "partition state reset");
```

Comment:
I don't think we need coverage this fine-grained here; we get plenty of guarantees for the internal state tracking of the `PartitionState` class in the `testOffsetSync` and `testZeroOffsetSync` cases. Adding more here is going to make the test more brittle, and IMO the tradeoff isn't worth it.

All I'd want to see out of a test case like this is this series of events:
1. Cause offset sync to take place by invoking `commitRecord`
2. Invoke producer callback used for step 1 to signal to the task that the offset sync succeeded
3. Cause another offset sync to take place by invoking `commitRecord`
4. Do not invoke the producer callback used for step 3 (yet)
5. Cause another offset sync to take place, and verify that no calls to `Producer::send` were made as a result
6. Invoke the producer callback used for step 3
7. Cause another offset sync to take place, and verify that `Producer::send` was invoked as a result

We can optionally verify the actual content of the offset syncs (i.e., upstream offset/downstream offset/topic partition).

Other details like the assertions on the values of any `PartitionState` fields can and should be omitted.

Reply:
Updated.